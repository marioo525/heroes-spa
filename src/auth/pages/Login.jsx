import { useContext, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const Login = () => {

  const {login} = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {

    const lastPath = localStorage.getItem('lastPath') || '/';

    login('Mario Camacho');

    navigate(lastPath, {
      replace: true,
    });
  };

  useEffect( () => {
    document.body.classList.add("login-page");

    return () => {
      document.body.classList.remove("login-page");
    }
  }, []);

  return (
      <div className="container mt-5 text-center center">
        <h1 className="login fw-bold">Login</h1>

        <button className="btn btn-primary button fs-6" onClick={onLogin}>
          Login
        </button>
      </div>
  );
};
