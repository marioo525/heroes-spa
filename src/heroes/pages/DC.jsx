import { HeroList } from "../components"

export const DC = () => {
  return (
    <>
      <h1 className="titulos">DC Comics</h1>
      <hr className="titulos"/>

      <HeroList publisher='DC Comics'/>
    </>
  )
}
