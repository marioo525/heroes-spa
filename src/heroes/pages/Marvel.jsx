import { HeroList } from "../components"

export const Marvel = () => {
  return (
    <>
      <h1 className="titulos">Marvel Comics</h1>
      <hr className="titulos"/>

      <HeroList publisher={'Marvel Comics'}/>
    </>
  )
}
