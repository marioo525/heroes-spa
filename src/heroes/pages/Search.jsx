import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import queryString from 'query-string'
import { getHeroesByName } from "../helpers";

export const Search = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const {q = ''} = queryString.parse(location.search);
  const heroes = getHeroesByName(q);

  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heroes.length === 0;

  const { searchText, onInputChange } = useForm({
    searchText: q
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    // if(searchText.trim().length <= 1) return;

    navigate(`?q=${ searchText.toLowerCase().trim() }`)
  }

  return (
    <>
      <h1 className="titulos">Búsqueda</h1>
      <hr className="titulos"/>

      <div className="row">
        <div className="col-5">
          <h4 className="titulos">Buscando</h4>
          <hr className="titulos"/>
          <form onSubmit={ onSearchSubmit } aria-label="form">
            <input
              type="text"
              placeholder="Busca un héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }
            />

            <button className="btn btn-primary mt-2">Buscar</button>
          </form>
        </div>

        <div className="col-7">
          <h4 className="titulos">Resultados</h4>
          <hr className="titulos"/>

          {/* {
            ( q === '' )
            ? <div className="alert alert-primary">Busca un héroe</div>
            : ( heroes.length === 0) 
              && <div className="alert alert-danger">No hay resultados con <b>{ q }</b></div>
          } */}

        <div className="alert alert-primary animate__animated animate__fadeIn" style={{ display: showSearch ? '' : 'none'}}>Busca un héroe</div>

        <div aria-label="alert-danger" className="alert alert-danger animate__animated animate__fadeIn" style={{ display: showError ? '' : 'none'}}>No hay resultados con <b>{ q }</b></div>

          {
            heroes.map( hero => (
              <HeroCard key={hero.id} {...hero} />
            ))
          }

        </div>
      </div>
    </>
  );
};
