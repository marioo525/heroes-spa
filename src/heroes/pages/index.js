export * from './Marvel';
export * from './DC';
export * from './Search';
export * from './Hero';