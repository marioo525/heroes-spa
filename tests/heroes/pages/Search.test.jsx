import { Search } from "../../../src/heroes/pages/Search";

import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

describe("Pruebas en <Search />", () => {

    beforeEach(() => jest.clearAllMocks());

  test("Debe de mostrarse correctamente con valores por defecto", () => {

    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();

  });

  test('Debe de mostrar a Batman y el input con el valor del queryString', () => {

    render(
        <MemoryRouter initialEntries={['/search?q=batman']}>
          <Search />
        </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');
    const img = screen.getByRole('img');
    expect(img.src).toContain('/heroes/dc-batman.jpg');

    const alert = screen.getByLabelText('alert-danger');    
    expect(alert.style.display).toBe('none');

  });

  test('Debe de mostrar un error si no se encuentra el hero (batman123)', () => {

    render(
        <MemoryRouter initialEntries={['/search?q=batman123']}>
          <Search />
        </MemoryRouter>
    );
    
    const alert = screen.getByLabelText('alert-danger');    
    expect(alert.style.display).toBe('');

  });

  test('Debe de llamar el navigate a la pantalla nueva', () => {

    render(
        <MemoryRouter initialEntries={['/search']}>
          <Search />
        </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {target: {name: 'searchText',value: 'superman'}})
    
    const form = screen.getByRole('form');
    fireEvent.submit(form);
    expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman');

  });


});
