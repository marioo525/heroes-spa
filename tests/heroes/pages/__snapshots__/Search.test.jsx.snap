// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    class="titulos"
  >
    Búsqueda
  </h1>
  <hr
    class="titulos"
  />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4
        class="titulos"
      >
        Buscando
      </h4>
      <hr
        class="titulos"
      />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Busca un héroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-2"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4
        class="titulos"
      >
        Resultados
      </h4>
      <hr
        class="titulos"
      />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Busca un héroe
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hay resultados con 
        <b />
      </div>
    </div>
  </div>
</div>
`;
